src/Pate/Proof.hs

```diff
@@ -353,296 +378,162 @@ traverseProofApp f = \case
+  ProofExpr sym arch tp ->
+  m (ProofExpr sym arch tp)
 traverseProofExpr f e =
   ProofExpr <$> (traverseProofApp (traverseProofExpr f) =<< (unApp <$> f e))
 
-transformProofExpr ::
-  Monad m =>
-  ProofTransformer m prf prf' ->
-  ProofExpr prf tp ->
-  m (ProofExpr prf' tp)
-transformProofExpr f (ProofExpr app) =
-  ProofExpr <$> (traverseProofApp (transformProofExpr f) =<< (transformProofApp f app))
-
 -- | Collect results from all sub-expressions
 collectProofExpr ::
-  forall w prf tp.
+  forall w sym arch tp.
   Monoid w =>
-  (forall tp'. ProofExpr prf tp' -> w) ->
-  ProofExpr prf tp ->
+  (forall tp'. ProofExpr sym arch tp' -> w) ->
+  ProofExpr sym arch tp ->
   w
 collectProofExpr f e_outer = runIdentity $ CMW.execWriterT (traverseProofExpr go e_outer)
   where
-    go :: ProofExpr prf tp' -> CMW.WriterT w Identity (ProofExpr prf tp')
+    go :: ProofExpr sym arch tp' -> CMW.WriterT w Identity (ProofExpr sym arch tp')
     go e = do
       CMW.tell (f e)
       return e
 
--- | A domain expression is actually independent of the 'app' parameter, so we can
--- project it to any one.
-appDomain ::
-  ProofExpr prf ProofDomainType ->
-  ProofApp prf app ProofDomainType
-appDomain (ProofExpr (ProofDomain regs stack glob domCtx)) = ProofDomain regs stack glob domCtx
-
-emptyDomain ::
-  forall prf m.
-  Monad m =>
-  IsProof prf =>
-  IsBoolLike prf m =>
-  ProofContext prf ->
-  m (ProofExpr prf ProofDomainType)
-emptyDomain domCtx = do
-  regs <- MM.mkRegStateM (\_ -> Const <$> proofPredFalse @prf)
-  stack <- emptyMemDomain
-  globs <- emptyMemDomain
-  return $ ProofExpr $ ProofDomain regs stack globs domCtx
-
-emptyMemDomain ::
-  forall prf m.
-  Monad m =>
-  IsBoolLike prf m =>
-  m (ProofMemoryDomain prf)
-emptyMemDomain = do
-  falsePred <- proofPredFalse @prf
-  return $ ProofMemoryDomain MapF.empty falsePred
-
-type family ProofScope prf :: DK.Type
+type family SymScope sym :: DK.Type
+type instance SymScope (W4B.ExprBuilder t fs scope) = t
 
 -- | A nonce representing an indirect reference to a proof node.
-data ProofNonce prf (tp :: ProofNodeType) where
-  ProofNonce :: N.Nonce (ProofScope prf) tp -> ProofNonce prf tp
+data ProofNonce sym (tp :: ProofNodeType) where
+  ProofNonce :: N.Nonce (SymScope sym) tp -> ProofNonce sym tp
 
-deriving instance Show (ProofNonce prf tp)
+deriving instance Show (ProofNonce sym tp)
 
-instance ShowF (ProofNonce prf) where
+instance ShowF (ProofNonce sym) where
   showF (ProofNonce n) = showF n
 
-instance TestEquality (ProofNonce prf) where
+instance TestEquality (ProofNonce sym) where
   testEquality (ProofNonce n1) (ProofNonce n2) = testEquality n1 n2
 
 instance OrdF (ProofNonce prf) where
   compareF (ProofNonce n1) (ProofNonce n2) = compareF n1 n2
 
-proofNonceValue :: ProofNonce prf tp -> Natural
+proofNonceValue :: ProofNonce sym tp -> Natural
 proofNonceValue (ProofNonce n) = fromIntegral (N.indexValue n)
 
 -- | A proof expression, annotated with nonces.
```

Comment:
It isn't just that it adds nonces - the `ProofNonceExpr` encodes the tree structure of the proof with its parent pointers

Reply:
Expanded on this comment a bit